README.md
[![shield to Libra's guild](https://discordapp.com/api/guilds/538054682699104256/embed.png?style=shield)](https://discord.gg/UuVHSZR)


# :bulb: www.Libragar.pw :bulb:
<b>:bulb: Hi there. Here I will tell you how I created my own server. :) :bulb:</b>

Original repository -> 

https://github.com/Adil950/OgarMulti -> for more information-> /OgarMulti/blob/master/README.md 

## Running
Video Tutorial [![Watch the video](https://i.imgur.com/RHqwwEU.png)](https://youtube.com/libraa)


### 0 - To be done first!
- buy a domain. //@godaddy
- buy a Vps. / Ubuntu 16.04 , 18.04 or 20.04. //@veritescil, @pulseheberg, @codeanywhere, @contabo(not test)
- buy a web hosting. (free cpanel) //@odeaweb
- replace your website's NS names with the NS names of your webhosting
- cpanel zone editor, add tip "A" your "vps ip" (noport).  "yoursubdomain.yourdomain.com" 



 ### 1 - Vps General installation


`sudo apt-get update`

`sudo apt-get install nodejs`


`sudo apt-get install nano`


`sudo apt-get install npm`

`apt install npm`



`apt-get update`

`apt-get install git`

`git --version`


### 2 - Nginx install 

`sudo apt-get update`


`sudo apt-get upgrade`


`sudo apt-get install nginx`


`sudo systemctl start nginx`

`sudo systemctl enable nginx`


`systemctl status nginx `

`nginx -t `

`rm /etc/nginx/sites-enabled/default`






### 3 - <b> Nginx ffa Configirasyon</b> 

`nano /etc/nginx/sites-available/ffa.conf` 

// after replacing it with your own information, put the following code into it.

	server {
		listen 80;
		listen [::]:80;

		server_name yoursubdomain.yourdomain.com;

		location / {
		proxy_pass http://yourserverip:3000;
		proxy_http_version 1.1;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection "upgrade";
		}
	}


save -> <b>ctrl + o</b> / <b>enter</b> /<b> ctrl + x</b>

 
`ln -s /etc/nginx/sites-available/ffa.conf /etc/nginx/sites-enabled/`

`nginx -t`

`systemctl reload nginx` OR `systemctl restart nginx` 



### 4 - Git clone and Installer.

`mkdir ffa`

`cd ffa`

`git clone https://github.com/Libra-Agario/Libragar`

`cd Libragar`

`sudo apt install nodejs`

`node -v`

`npm install ws`

`sudo apt update`

`sudo apt install build-essential`

`sudo apt-get install manpages-dev`

`gcc --version`

`sudo apt-get install nano`

`npm install pm2@latest -g`


### 5 - SSL Certifica

`apt install certbot python3-certbot-nginx`  # OR # `sudo apt install certbot python3-certbot-nginx`


`sudo certbot --nginx -d yoursubdomain.yourdomain.com --register-unsafely-without-email`

`A`

`2`

`sudo certbot renew --dry-run`



### 6 - pm2 server start -> 

`cd cli`

`pm2 list`

`pm2 start index.js`

`pm2 list`

`pm2 save`



`pm2 unstartup` 

`pm2 startup`

`pm2 save`


### 7 - Libragar Script + Tampermonkey Install -> 
1. https://www.libragar.pw OR /Libragar/User-Scripts (https://github.com/Libra-Agario/Libragar/tree/main/User-Scripts
OR https://greasyfork.org/en/scripts/420590-www-libragar-pw-libra-agario-yt-public-betatest-server 

### 8.1 Libragar Script in edit ->
1. Libragar Script -> in Search -> `core.disableIntegrityChecks(true); core.connect` replace the same ones with your own link.
 - <b> Sample -> </b> core.disableIntegrityChecks(true); core.connect("wss://`ffa.libragar.pw`");
 - <b> Change -> </b>  core.disableIntegrityChecks(true); core.connect("wss://`yoursubdomain.yourdomain.com`");


### 8 - </b> Commands you can do via pm2 ssh.

For the feedback -> `pm2 log --lines n` (CTRL+C = Log Exit)

Sample 1 ->  `pm2 send 0 'help'`  &   /Libragar/CommandList.txt

Sample 2 ->  `pm2 send 0 'addminion 1 100'` 

Other pm2 commands -> `pm2 --help`




### 9 To create more servers - (not mandatory)
follow the procedures here. -> [Add More Servers](https://github.com/Libra-Agario/Libragar/blob/main/LibraOtherDocs/addmoreservers.md)


## Configuring

- To change how Libragar runs, modify `cli/settings.json`. // Game mode and in-game settings.

 Ask all your questions over on the [Agar.io Private Servers](https://discord.gg/66X2ESb) Discord guild.


CHANGELOG
1.3.7 @ 10.05.2021 / Libra Agario YT
- New Setting, userFrozen the user can now freeze himself by the server with the help of a key.
- Fixed FFA self- crash issue. (seems to be working well for now) 
- it was simply upgraded to protocol 23. it supports all protocols Realy agar.io
- it was replaced by the uws - ws as it was constantly in trouble. change it on demand.

1.3.6 @ 2019-11-02
- Support legacy protocol 22
- New setting, worldEatMult
- New setting, worldEatOverlapDiv
- New setting, playerSplitSizeDiv
- New setting, playerSplitDistance
- Use size instead of mass for merge time calculation

1.3.5 @ 2019-07-29
- Add message length checks to modern protocol
- Move sqrt(2) and sqrt(1.3) into constants in Misc.js
- Fill JSDoc for context argument in default commands
- Formatting

1.3.4 @ 2019-07-13
- Fix autosplitting never wanting to split beyond max - 1
- Setting negative size to a cell will now immediately cause crash

1.3.3 @ 2019-07-09
- Implement legacy protocol 21's different Pie leaderboard message netcode
- POTENTIALLY BREAKING: Define Teams gamemode's type as 2 instead of 1, to avoid agar.io registering it as the new Team mode with royale-like properties
- Language refactoring

1.3.2 @ 2019-07-06
- Fix crash where a connection with no protocol gets added into global chat and gets sent a message
- Fix server message having a black color
- Rename definition interface Range to Rect to avoid confusion with browser definition of same name

1.3.1 @ 2019-06-26
- Fix Q presses not registering for legacy protocol
- When equal size playercells rigid collide, the distance traveled of both is half the size; retroactively apply this to bot splitting

1.3.0 @ 2019-06-16
- Fix command killall improperly querying for players in world
- Fix command setting being susceptible of code injections
- Fix inconsistency for server load on modern protocol
- New setting, worldPlayerBotSkins, which will be used if a bot's skin in name is *
- New setting, playerAllowSkinInName, which will allow players to set skin via name where enclosed in <>
- Colors are internally defined as number hex instead of an interface
- Decrease lowlight / highlight difference in Teams
- Streamline long name trimming
- Formatting

1.2.5 @ 2019-06-05
- Fix crash when executing worldid player command
- Fix joinworld player command not letting player with no world join

1.2.4 @ 2019-06-03
- Support legacy protocol 21
- Support modern protocol 3
- Fix playercells staying in one place when split at the same tick
- Fix free-roaming not working
- New setting, worldMinCount, prevents empty worlds from destroying
- Updated setting ejectedSize to match vanilla
- Updated setting ejectingLoss to match vanilla
- Format PlayerBot.canSplitkill
- Format Ticker
- If setting minionName is "*", the minion spawning name will include the following player's name
- Minions will not spawn if the following player is not alive
- Replace QuadTree.getItems to getItemCount as getting all items is useless
- Use util.inspect for eval command
- Routers now close when the server handle closes

1.2.3 @ 2019-04-26
- Fix crash when executing stats command
- Fix bad control flow when handle isn't running and executing stats command
- New setting, playerSplitCap, to control maximum amount of possible splits in one tick
- New setting, playerEjectDelay, to control ejecting speed

1.2.2 @ 2019-03-27
- Fix crash on ping or pong websocket event due to bad bind
- Moved Router from primitives to sockets
- Removed settings listenerMinLegacyProtocol and listenerMaxLegacyProtocol
- Updated command eval to show error stack on thrown error
- New command, crash
- Updated JSDoc for QuadTree for template support
- Minor bug fixes with log handler

1.2.1 @ 2019-03-25
- Clean up gitignore
- Add missing JSDoc to CLI log handler
- Renamed Ticker.isRunning to running
- Renamed definition Position to Point
- Refactored CommandList JSDoc for better type recognition in default commands
- Refactored default commands
- Refactored usage of first player cell for player name and color
- Updated stats command
- Moved registering gamemodes, protocols and commands to be done in CLI
- Fixed bug where internal routers' ACTIVE and DORMANT columns in routers command are NaN
- Removed command pause
- Removed unnecessary requires
- Map IPv6-mapped IPv4 remote addresses to true IPv4
- Check connections for forbidden IPs in Listener update
- New command, forbid
- New command, pardon
- Renamed setting serverChatEnabled to chatEnabled
- New setting, chatCooldown
- New setting, chatFilteredPhrases
- Prepare for NPM release

1.2.0 @ 2019-02-17
- Fix crash when leaving world
- Fix crash when spectating a player that gets destroyed
- Upon exploding the split angles are now random
- Support legacy protocol 20
- SIGINT on command stream will close the server. Two SIGINTs for this are not needed anymore
- Implement worldPlayerBotNames setting
- Renamed setting serverUpdateFrequency to serverFrequency
- New command, players
- Updated routers command
- Updated setting command to Suggest setting names if given doesn't exist
- Renamed command killminion to rmminion
- Renamed command killbot to rmbot
- Formatting and language changes

1.1.12 @ 2018-11-12
- Fix crash when leaving world then using worldid chat command
- Implement modern protocol revision 2
- Client gets its visible cells and leaderboard cleared when leaving a world
- Formatting

1.1.11 @ 2018-08-31
- Fix wrong usage of unsigned integers for cell position in legacy protocol
- Fix crash when matchmakerNeedsQueuing is on
- Fix virus count being able to exceed virusMaxCount

src/protocols/LegacyProtocol.js
const Protocol = require("./Protocol");
const Reader = require("../primitives/Reader");
const Writer = require("../primitives/Writer");

class LegacyProtocol extends Protocol {
    /**
     * @param {Connection} connection
     */
    constructor(connection) {
        super(connection);
        this.gotProtocol = false;
        this.protocol = NaN;
        this.gotKey = false;
        this.key = NaN;

        this.lastLeaderboardType = null;
    }

    static get type() { return "legacy"; }
    get subtype() { return `l${!isNaN(this.protocol) ? ("00" + this.protocol).slice(-2) : "//"}`; }

    /**
     * @param {Reader} reader
     */
    distinguishes(reader) {
        if (reader.length < 5) return false;
        if (reader.readUInt8() !== 254) return false;
        this.gotProtocol = true;
        this.protocol = reader.readUInt32();
        if (this.protocol < 4) {
            this.protocol = 4;
            this.logger.debug(`legacy protocol: got version ${this.protocol}, which is lower than 4`);
        }
        return true;
    }

    /**
     * @param {Reader} reader
     */
    onSocketMessage(reader) {
        const messageId = reader.readUInt8();
        if (!this.gotKey) {
            if (messageId !== 255) return;
            if (reader.length < 5) return void this.fail("Unexpected message format");
            this.gotKey = true;
            this.key = reader.readUInt32();
            this.connection.createPlayer();
            return;
        }
        switch (messageId) {
            case 0:
                this.connection.spawningName = readZTString(reader, this.protocol);
                break;
            case 1:
                this.connection.requestingSpectate = true;
                break;
            case 16:
                switch (reader.length) {
                    case 13:
                        this.connection.mouseX = reader.readInt32();
                        this.connection.mouseY = reader.readInt32();
                        break;
                    case 9:
                        this.connection.mouseX = reader.readInt16();
                        this.connection.mouseY = reader.readInt16();
                        break;
                    case 21:
                        this.connection.mouseX = ~~reader.readFloat64();
                        this.connection.mouseY = ~~reader.readFloat64();
                        break;
                    default: return void this.fail(1003, "Unexpected message format");
                }
                break;
            case 17:
                if (this.connection.controllingMinions)
                    for (let i = 0, l = this.connection.minions.length; i < l; i++)
                        this.connection.minions[i].splitAttempts++;
                else this.connection.splitAttempts++;
                break;
            case 18: this.connection.isPressingQ = true; break;
            case 19: this.connection.isPressingQ = this.hasProcessedQ = false; break;
            case 21:
                if (this.connection.controllingMinions)
                    for (let i = 0, l = this.connection.minions.length; i < l; i++)
                        this.connection.minions[i].ejectAttempts++;
                else this.connection.ejectAttempts++;
                break;
            case 22:
                if (!this.gotKey || !this.settings.minionEnableERTPControls) break;
                for (let i = 0, l = this.connection.minions.length; i < l; i++)
                    this.connection.minions[i].splitAttempts++;
                break;
            case 23:
                if (!this.gotKey || !this.settings.minionEnableERTPControls) break;
                for (let i = 0, l = this.connection.minions.length; i < l; i++)
                    this.connection.minions[i].ejectAttempts++;
                break;
            case 24:
                if (!this.gotKey || !this.settings.minionEnableERTPControls) break;
                this.connection.minionsFrozen = !this.connection.minionsFrozen;
                break;				
	        case 25: // Libra edit 1 
                if (!this.gotKey || !this.settings.minionEnableERTPControls) break;
                this.connection.userFrozen = !this.connection.userFrozen;
                break;
				
            case 99:
                if (reader.length < 2)
                    return void this.fail(1003, "Bad message format");
                const flags = reader.readUInt8();
                const skipLen = 2 * ((flags & 2) + (flags & 4) + (flags & 8))
                if (reader.length < 2 + skipLen)
                    return void this.fail(1003, "Unexpected message format");
                reader.skip(skipLen);
                const message = readZTString(reader, this.protocol);
                this.connection.onChatMessage(message);
                break;
            case 254:
                if (this.connection.hasPlayer && this.connection.player.hasWorld)
                    this.onStatsRequest();
                break;
            case 255: return void this.fail(1003, "Unexpected message");
            default: return void this.fail(1003, "Unknown message type");
        }
    }

    /**
     * @param {ChatSource} source
     * @param {string} message
     */
    onChatMessage(source, message) {
        const writer = new Writer();
        writer.writeUInt8(99);
        writer.writeUInt8(source.isServer * 128);
        writer.writeColor(source.color);
        writeZTString(writer, source.name, this.protocol);
        writeZTString(writer, message, this.protocol);
        this.send(writer.finalize());
    }

    onStatsRequest() {
        const writer = new Writer();
        writer.writeUInt8(254);
        const stats = this.connection.player.world.stats;
        const legacy = {
            mode: stats.gamemode,
            update: stats.loadTime,
            playersTotal: stats.external,
            playersAlive: stats.playing,
            playersSpect: stats.spectating,
            playersLimit: stats.limit
        };
        writeZTString(writer, JSON.stringify(Object.assign({}, legacy, stats)), this.protocol);
        this.send(writer.finalize());
    }

    /**
     * @param {PlayerCell} cell
     */
    onNewOwnedCell(cell) {
        const writer = new Writer();
        writer.writeUInt8(32);
        writer.writeUInt32(cell.id);
        this.send(writer.finalize());
    }

    /**
     * @param {Rect} range
     * @param {boolean} includeServerInfo
     */
    onNewWorldBounds(range, includeServerInfo) {
        const writer = new Writer();
        writer.writeUInt8(64);
        writer.writeFloat64(range.x - range.w);
        writer.writeFloat64(range.y - range.h);
        writer.writeFloat64(range.x + range.w);
        writer.writeFloat64(range.y + range.h);
        if (includeServerInfo) {
            writer.writeUInt32(this.handle.gamemode.type);
            writeZTString(writer, `OgarII ${this.handle.version}`, this.protocol);
        }
        this.send(writer.finalize());
    }

    onWorldReset() {
        const writer = new Writer();
        writer.writeUInt8(18);
        this.send(writer.finalize());
        if (this.lastLeaderboardType !== null) {
            this.onLeaderboardUpdate(this.lastLeaderboardType, [], null);
            this.lastLeaderboardType = null;
        }
    }

    /**
     * @param {LeaderboardType} type
     * @param {LeaderboardDataType[type][]} data
     * @param {LeaderboardDataType[type]=} selfData
     */
    onLeaderboardUpdate(type, data, selfData) {
        this.lastLeaderboardType = type;
        const writer = new Writer();
        switch (type) {
            case "ffa": ffaLeaderboard[this.protocol](writer, data, selfData, this.protocol); break;
            case "pie": pieLeaderboard[this.protocol](writer, data, selfData, this.protocol); break;
            case "text": textBoard[this.protocol](writer, data, this.protocol); break;
        }
        this.send(writer.finalize());
    }

    /**
     * @param {ViewArea} viewArea
     */
    onSpectatePosition(viewArea) {
        const writer = new Writer();
        writer.writeUInt8(17);
        writer.writeFloat32(viewArea.x);
        writer.writeFloat32(viewArea.y);
        writer.writeFloat32(viewArea.s);
        this.send(writer.finalize());
    }

    /**
     * @abstract
     * @param {Cell[]} add
     * @param {Cell[]} upd
     * @param {Cell[]} eat
     * @param {Cell[]} del
     */
    onVisibleCellUpdate(add, upd, eat, del) {
        const source = this.connection.player;
        const writer = new Writer();
        writer.writeUInt8(16);
        let i, l, cell;

        l = eat.length;
        writer.writeUInt16(l);
        for (i = 0; i < l; i++) {
            cell = eat[i];
            writer.writeUInt32(cell.eatenBy.id);
            writer.writeUInt32(cell.id);
        }

        for (i = 0, l = add.length; i < l; i++) {
            cell = add[i];
            writeCellData[this.protocol](writer, source, this.protocol, cell,
                true, true, true, true, true, true);
        }
        for (i = 0, l = upd.length; i < l; i++) {
            cell = upd[i];
            writeCellData[this.protocol](writer, source, this.protocol, cell,
                false, cell.sizeChanged, cell.posChanged, cell.colorChanged, cell.nameChanged, cell.skinChanged);
        }
        writer.writeUInt32(0);

        l = del.length;
        writer[this.protocol < 6 ? "writeUInt32" : "writeUInt16"](l);
        for (i = 0; i < l; i++) writer.writeUInt32(del[i].id);
        this.send(writer.finalize());
    }
}

module.exports = LegacyProtocol;

/**
 * @type {{ [protocol: number]: (writer: Writer, data: LeaderboardDataType["pie"][], selfData: LeaderboardDataType["pie"], protocol: number) => void }}
 */
const pieLeaderboard = {
    4: pieLeaderboard4,
    5: pieLeaderboard4,
    6: pieLeaderboard4,
    7: pieLeaderboard4,
    8: pieLeaderboard4,
    9: pieLeaderboard4,
    10: pieLeaderboard4,
    11: pieLeaderboard4,
    12: pieLeaderboard4,
    13: pieLeaderboard4,
    14: pieLeaderboard4,
    15: pieLeaderboard4,
    16: pieLeaderboard4,
    17: pieLeaderboard4,
    18: pieLeaderboard4,
    19: pieLeaderboard4,
    20: pieLeaderboard4,
    21: pieLeaderboard21,
    22: pieLeaderboard21,
    23: pieLeaderboard21 // Libra edit 4
};
/**
 * @param {Writer} writer
 * @param {LeaderboardDataType["pie"][]} data
 * @param {LeaderboardDataType["pie"]=} selfData
 * @param {number} protocol
 */
function pieLeaderboard4(writer, data, protocol) {
    writer.writeUInt8(50);
    writer.writeUInt32(data.length);
    for (let i = 0, l = data.length; i < l; i++)
        writer.writeFloat32(data[i].weight);
}
/**
 * @param {Writer} writer
 * @param {LeaderboardDataType["pie"][]} data
 * @param {LeaderboardDataType["pie"]=} selfData
 * @param {number} protocol
 */
function pieLeaderboard21(writer, data, protocol) {
    writer.writeUInt8(50);
    writer.writeUInt32(data.length);
    for (let i = 0, l = data.length; i < l; i++) {
        writer.writeFloat32(data[i].weight);
        writer.writeColor(data[i].color);
    }
}

/**
 * @type {{ [protocol: number]: (writer: Writer, data: LeaderboardDataType["ffa"][], selfData: LeaderboardDataType["ffa"], protocol: number) => void }}
 */
const ffaLeaderboard = {
    4: ffaLeaderboard4,
    5: ffaLeaderboard4,
    6: ffaLeaderboard4,
    7: ffaLeaderboard4,
    8: ffaLeaderboard4,
    9: ffaLeaderboard4,
    10: ffaLeaderboard4,
    11: ffaLeaderboard11,
    12: ffaLeaderboard11,
    13: ffaLeaderboard11,
    14: ffaLeaderboard11,
    15: ffaLeaderboard11,
    16: ffaLeaderboard11,
    17: ffaLeaderboard11,
    18: ffaLeaderboard11,
    19: ffaLeaderboard11,
    20: ffaLeaderboard11,
    21: ffaLeaderboard11,
    22: ffaLeaderboard11,
    23: ffaLeaderboard11 // Libra edit 4
};
/**
 * @param {Writer} writer
 * @param {LeaderboardDataType["ffa"][]} data
 * @param {LeaderboardDataType["ffa"]=} selfData
 * @param {number} protocol
 */
function ffaLeaderboard4(writer, data, selfData, protocol) {
    writer.writeUInt8(49);
    writer.writeUInt32(data.length);
    for (let i = 0, l = data.length; i < l; i++) {
        const item = data[i];
        if (protocol === 6)
            writer.writeUInt32(item.highlighted ? 1 : 0);
        else writer.writeUInt32(item.cellId);
        writeZTString(writer, item.name, protocol);
    }
}
/**
 * @param {Writer} writer
 * @param {LeaderboardDataType["ffa"][]} data
 * @param {LeaderboardDataType["ffa"]=} selfData
 * @param {number} protocol
 */
function ffaLeaderboard11(writer, data, selfData, protocol) {
    writer.writeUInt8(protocol >= 14 ? 53 : 51);
    for (let i = 0, l = data.length; i < l; i++) {
        const item = data[i];
        if (item === selfData)
            writer.writeUInt8(8);
        else {
            writer.writeUInt8(2);
            writer.writeZTStringUTF8(item.name);
        }
    }
}

/**
 * @type {{ [protocol: number]: (writer: Writer, data: LeaderboardDataType["text"][], protocol: number) => void }}
 */
const textBoard = {
    4: textBoard4,
    5: textBoard4,
    6: textBoard4,
    7: textBoard4,
    8: textBoard4,
    9: textBoard4,
    10: textBoard4,
    11: textBoard4,
    12: textBoard4,
    13: textBoard4,
    14: textBoard14,
    15: textBoard14,
    16: textBoard14,
    17: textBoard14,
    18: textBoard14,
    19: textBoard14,
    20: textBoard14,
    21: textBoard14,
    22: textBoard14,
    23: textBoard14 // Libra edit 4
};
/**
 * @param {Writer} writer
 * @param {LeaderboardDataType["text"][]} data
 * @param {number} protocol
 */
function textBoard4(writer, data, protocol) {
    writer.writeUInt8(48);
    writer.writeUInt32(data.length);
    for (let i = 0, l = data.length; i < l; i++)
        writer.writeZTString(data[i], protocol);
}
/**
 * @param {Writer} writer
 * @param {LeaderboardDataType["text"][]} data
 * @param {number} protocol
 */
function textBoard14(writer, data, protocol) {
    writer.writeUInt8(53);
    for (let i = 0, l = data.length; i < l; i++) {
        writer.writeUInt8(2);
        writer.writeZTStringUTF8(data[i]);
    }
}

/**
 * @type {{ [protocol: number]: (writer: Writer, source: Player, protocol: number, cell: Cell, includeType: boolean, includeSize: boolean, includePos: boolean, includeColor: boolean, includeName: boolean, includeSkin: boolean) => void }}
 */
const writeCellData = {
    4: writeCellData4,
    5: writeCellData4,
    6: writeCellData6,
    7: writeCellData6,
    8: writeCellData6,
    9: writeCellData6,
    10: writeCellData6,
    11: writeCellData11,
    12: writeCellData11,
    13: writeCellData11,
    14: writeCellData11,
    15: writeCellData11,
    16: writeCellData11,
    17: writeCellData11,
    18: writeCellData11,
    19: writeCellData11,
    20: writeCellData11,
    21: writeCellData11,
    22: writeCellData11,
    23: writeCellData11 // Libra edit 4
};
/**
 * @param {Writer} writer
 * @param {Player} source
 * @param {number} protocol
 * @param {Cell} cell
 * @param {boolean} includeType
 * @param {boolean} includeSize
 * @param {boolean} includePos
 * @param {boolean} includeColor
 * @param {boolean} includeName
 * @param {boolean} includeSkin
 */
function writeCellData4(writer, source, protocol, cell, includeType, includeSize, includePos, includeColor, includeName, includeSkin) {
    writer.writeUInt32(cell.id);
    writer[protocol === 4 ? "writeInt16" : "writeInt32"](cell.x);
    writer[protocol === 4 ? "writeInt16" : "writeInt32"](cell.y);
    writer.writeUInt16(cell.size);
    writer.writeColor(cell.color);

    let flags = 0;
    if (cell.isSpiked) flags |= 0x01;
    if (includeSkin) flags |= 0x04;
    if (cell.isAgitated) flags |= 0x10;
    if (cell.type === 3) flags |= 0x20;
    writer.writeUInt8(flags);

    if (includeSkin) writer.writeZTStringUTF8(cell.skin);
    if (includeName) writer.writeZTStringUCS2(cell.name);
    else writer.writeUInt16(0);
}
/**
 * @param {Writer} writer
 * @param {Player} source
 * @param {number} protocol
 * @param {Cell} cell
 * @param {boolean} includeType
 * @param {boolean} includeSize
 * @param {boolean} includePos
 * @param {boolean} includeColor
 * @param {boolean} includeName
 * @param {boolean} includeSkin
 */
function writeCellData6(writer, source, protocol, cell, includeType, includeSize, includePos, includeColor, includeName, includeSkin) {
    writer.writeUInt32(cell.id);
    writer.writeInt32(cell.x);
    writer.writeInt32(cell.y);
    writer.writeUInt16(cell.size);

    let flags = 0;
    if (cell.isSpiked) flags |= 0x01;
    if (includeColor) flags |= 0x02;
    if (includeSkin) flags |= 0x04;
    if (includeName) flags |= 0x08;
    if (cell.isAgitated) flags |= 0x10;
    if (cell.type === 3) flags |= 0x20;
    writer.writeUInt8(flags);

    if (includeColor) writer.writeColor(cell.color);
    if (includeSkin) writer.writeZTStringUTF8(cell.skin);
    if (includeName) writer.writeZTStringUTF8(cell.name);
}
/**
 * @param {Writer} writer
 * @param {Player} source
 * @param {number} protocol
 * @param {Cell} cell
 * @param {boolean} includeType
 * @param {boolean} includeSize
 * @param {boolean} includePos
 * @param {boolean} includeColor
 * @param {boolean} includeName
 * @param {boolean} includeSkin
 */
function writeCellData11(writer, source, protocol, cell, includeType, includeSize, includePos, includeColor, includeName, includeSkin) {
    writer.writeUInt32(cell.id);
    writer.writeInt32(cell.x);
    writer.writeInt32(cell.y);
    writer.writeUInt16(cell.size);

    let flags = 0;
    if (cell.isSpiked) flags |= 0x01;
    if (includeColor) flags |= 0x02;
    if (includeSkin) flags |= 0x04;
    if (includeName) flags |= 0x08;
    if (cell.isAgitated) flags |= 0x10;
    if (cell.type === 3) flags |= 0x20;
    if (cell.type === 3 && cell.owner !== source) flags |= 0x40;
    if (includeType && cell.type === 1) flags |= 0x80;
    writer.writeUInt8(flags);
    if (includeType && cell.type === 1) writer.writeUInt8(1);

    if (includeColor) writer.writeColor(cell.color);
    if (includeSkin) writer.writeZTStringUTF8(cell.skin);
    if (includeName) writer.writeZTStringUTF8(cell.name);
}

/**
 * @param {Reader} reader
 * @param {number} protocol
 * @returns {string}
 */
function readZTString(reader, protocol) {
    return reader[protocol < 6 ? "readZTStringUCS2" : "readZTStringUTF8"]();
}
/**
 * @param {Writer} writer
 * @param {string} value
 * @param {number} protocol
 */
function writeZTString(writer, value, protocol) {
    writer[protocol < 6 ? "writeZTStringUCS2" : "writeZTStringUTF8"](value);
}

const Cell = require("../cells/Cell");
const Player = require("../worlds/Player");
const PlayerCell = require("../cells/PlayerCell");
const Connection = require("../sockets/Connection");
